doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,222 @@
+- When an error occurred, the kind error can be distinguished after the fact in many cases.
+- Callers normally don't care about the exact error, unless they can do something about it.
+  This could be:
+  - reporting the error to a user
+  - switching to a different code path
+  - (during debugging) changing the code
+- Functions can be written in a such a way that they detect and report all errors without permanent effects.
+
+## Considered Alternatives
+
+### Extra argument
+
+All functions that report errors take a `int * error` argument to which errors can be written.
+
+This makes the API much more convoluted to use.
+
+A benefit would be the return value remains fully usable.
+
+### Thread Local Storage
+
+Use a field in thread local storage (TLS) to store errors similar to `errno`.
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Pragmatic solution
+
+The pragmatic solution here is to simply assume "everything will be fine".
+
+Many functions only return errors for simple precondition violations, like passing a `NULL` pointer or otherwise invalid argument.
+Especially for `libelektra-core` (the only place where `Key * errorKey` truly cannot be used), all functions must be side effect free.
+Therefore, the only possible cause of errors are invalid arguments.
+
+If we now assume the function is written such that it fails before it changes any argument's data, then we can infer that only a single error indicating return value is needed.
+Because the arguments are unchanged and the only possible cause of errors, we can distinguish between errors after the fact.
+
+As stated, many functions' only error cases are precondition violations.
+If the preconditions are simple enough, they can be checked beforehand if it is not clear whether the precondition holds.
+
+Take for example some hypothetical `keyValue` function, in a world where we have a `key->value` container for COW that should always be present.
+Looking just at the error checks, without even considering the function signature:
+
+```c
+if (key == NULL) {
+// error 1: precondition violation
+}
+
+if (key->value == NULL) {
+// error 2: invalid Key, COW container missing
+}
+
+// success: return value
+return key->value->data;
+```
+
+If we want separate error indications for both error cases and a nice API we immediately hit a problem.
+The obvious return type is `const void *`, because that's the type of the key value (and we want it const).
+But that's a pointer, so only `NULL` could be used for errors.
+The issue is, `NULL` may be a valid key value.
+So what do we do?
+
+Essentially, we do nothing, "everything is fine":
+
+```c
+const void * keyValue (Key * key) {
+  if (key == NULL) {
+    return NULL;
+  }
+
+  assert(key->value != NULL);
```

Comment:
You probably missed it and I probably should have used a different function name. In this hypothetical example, `key->value` is not the value itself, but the COW container. And also just in the hypothetical example (not in the real world) this should always be there.

Again, I wanted a very simple example to showcase this "internal precondition" case. I didn't use the name, because that cannot be `NULL`. So `NULL` is completely free to be used as an error value and doesn't clash (something I also wanted to show). Maybe I'll split it into multiple examples. I ended up with more examples than expected anyway...